Derive classes for rich enums

evonreis opened the pull request:
Add new attribute macro `gen_stub_pyclass_rich_enum` that creates internal classes for structured enum variants, and tuples for tuple variants.  

The internal classes reflect in Python how structured variants are exposed by PyO3.   

The tuple type hint may not be the correct way to handle tuple variants.

This change requires a new inventoried struct called `PyClassTreeInfo` that allows for nested classes.

`ClassDef ` now can be created from the import of either PyClassTreeInfo or PyClassInfo and will write indented internal classes to the stub file.

This PR also fixes the implementation on tuples so that tuples of length one are supported.

jagregory commented:
This looks great @evonreis. I just tried it out and it nearly works as I'd expected with a couple of small issues if you're interested in feedback.

1. The inner classes that PyO3 generates are subclassed from the outer class, but the typings in your PR don't reflect that.

   e.g. in your example it should be:

   ```python
   class NumberRich:
       class FLOAT(NumberRich): # <-- inherits from NumberRich
           r"""
           Float variant
           """
           _0: builtins.float
   
       class INTEGER(NumberRich):
           r"""
           Integer variant
           """
           int: builtins.int
           r"""
           The integer value
           """
   ```

   This way if you have a function elsewhere you can define it as `def something() -> NumberRich` allowing either branch of the enum to be returned. Currently this isn't possible as `FLOAT` and `INTEGER` are not typed as subclasses.

2. PyO3 generates default constructors for the classes which are missing from these typings, so currently it's not possible to construct these classes and make the type checker happy.

   ```python
   class NumberRich:
       class FLOAT:
           # ... existing code ...
           __new__(cls, _0: builtins.float): ... # <-- default constructor
   
       class INTEGER:
           # ... existing code ...
           __new__(cls, int: builtins.int): ...
   ```

evonreis commented:
Thank you @jagregory and @flying-sheep for looking this over.  I will address all of the issues you've brought up.

evonreis commented:
@jagregory, I've pushed a commit for the making the variant classes derived from the enum class in the stub file. 

I build a `TypeInfo` for the enum on the fly when building `ClassDef` structs.  This avoids having to specify the name of the enum in the `VariantInfo` struct, which in turn avoids a user having to enter the enum name for every variant if they are defining the stub by hand (not using the macro).

evonreis commented:
@jagregory I cherry-picked your commit for a variant `__new__()` method, and modified it to handle the `#[pyo3(constructor = ...)]` case.

`__match_case__`, `__len__()`, and `__getitem__()` are also added.

@jagregory I don't understand your testing suggestion, but certainly more tests are needed.  I think your other points are now addressed.

evonreis commented:
@flying-sheep I understand your idea better now.  I tried to write something, but assert_type() was introduced in py3.11 seemingly, and pyo3-stub-gen passes tests in py3.9.  Getting assert_type() to work while still supporting 3.9 is beyond my python knowledge.

flying-sheep commented:
No worries, the exact tests you added before are perfect!

jagregory commented:
What's remaining on this to be able to merge?

termoshtt commented:
Thanks PR, and sorry for long absence. I merged #204, and it causes conflict. Could you fix them?

evonreis commented:
Thanks for looking at this.  I've done the merge, ran tests and pushed the changes.

termoshtt reviewed (changes requested):
I've read `examples/` part and add some test based on [PyO3's document](https://pyo3.rs/v0.25.1/class.html#complex-enums). These works very well!

Before reading Rust part, I have a concern; it would be better to use "complex enum" rather than "rich enum". This feature is originated from PyO3, and PyO3 calls it "complex enum" in its document. If you do not have reason to use "rich enum", could you change it?

evonreis commented:
@termoshtt `rich` is now `complex`.
